### PR TITLE
Remove lowercase for the hostname as recommended/advised by OAuth spec

### LINKED ISSUE #25001: Client redirect_uri check must be compared using exact string matching
### Description

From Keycloak 21.1.2 as described in [this issue](https://github.com/keycloak/keycloak/issues/24792), Keycloak has a bug with redirect_uri with uppercase letters.
This was fixed in version 23.0.0 in a bad and insecure way with [this PR](https://github.com/keycloak/keycloak/pull/24819).

Maybe OAuth 2.0 allows case-insensitive redirect_uri check. Could you provide the specification part of it?
For sure, [OAuth 2.0 Security Best Current Practice](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics#name-protecting-redirect-based-f) mentions : 

> When comparing client redirect URIs against pre-registered URIs, authorization servers MUST utilize exact string matching except for port numbers in localhost redirection URIs of native apps, see [Section 4.1.3](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics#iuv_countermeasures). We believe Keycloak should follow OAuth 2.0 Security Best Current Practice. 

Moreover, this will be a step for implementing [OAuth 2.1](https://oauth.net/2.1/). : 

> The major differences from OAuth 2.0 are listed below.
>...
>- Redirect URIs must be compared using exact string matching

This issue could be a bug - or even a security issue, because it breaks [OAuth 2.0 Security Best Current Practice](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics#name-protecting-redirect-based-f).

### PR DESCRIPTION
**Pull request by rmartinc**
Closes https://github.com/keycloak/keycloak/issues/25001

Reverting  the previous PR https://github.com/keycloak/keycloak/pull/24819 ([commit](https://github.com/keycloak/keycloak/commit/1241bd291901667a13506d6fb123ba4344e618ee)) about lower case in `RedirectUtils` class. A new commit removes the lower case completely and modify/add tests to confirm we are using exact match for hostname and scheme. This PR also fixes the previous issue that the reverted commit was solving.

### CODE REVIEW

File: services/src/main/java/org/keycloak/protocol/oidc/utils/RedirectUtils.java [hunk 1]
```diff
@@ -72,8 +72,10 @@ public static Set<String> resolveValidRedirects(KeycloakSession session, String
             if (validRedirect.startsWith("/")) {
                 validRedirect = relativeToAbsoluteURI(session, rootUrl, validRedirect);
                 logger.debugv("replacing relative valid redirect with: {0}", validRedirect);
+                resolveValidRedirects.add(validRedirect);
+            } else {
+                resolveValidRedirects.add(validRedirect);
             }
-            resolveValidRedirects.add(lowerCaseHostname(validRedirect));
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/protocol/oidc/utils/RedirectUtils.java:76]: This a nitpick but as it was before (without the `else` block) read better, for me ...

Comment (rmartinc) [reply services/src/main/java/org/keycloak/protocol/oidc/utils/RedirectUtils.java:76]: :smile: Yep, also for me. I just reverted the previous commit and I didn't re-add that change. It's done now.

File: docs/documentation/upgrading/topics/keycloak/changes-23_0_2.adoc [hunk 2]
```diff
@@ -0,0 +1,3 @@
+= Valid redirect URIs for clients are always compared with exact string matching
+
+Version 1.8.0 introduced a lower-case for the hostname and scheme when comparing a redirect URI with the specified valid redirects for a client. Unfortunately it did not fully work in all the protocols, and, for example, the host was lower-cased for `http` but not for `https`. As https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics#name-protecting-redirect-based-f[OAuth 2.0 Security Best Current Practice] advises to compare URIs using exact string matching, {project_name} will follow the recommendation and for now on valid redirects are compared with exact case even for the hostname and scheme. Please be aware of the change if you were using insensitive case comparison in the redirect URIs.
```
[the remark this answers is not in the document]
Comment (rmartinc) [reply docs/documentation/upgrading/topics/keycloak/changes-23_0_2.adoc]: :+1: Done! (I removed the previous sentence `Please be aware of the...` then as it's better explained now.)

**Review by pedroigor — APPROVED** [review]

**Comment by pedroigor** [comment]
@mposolda wdyt?

**Review by mposolda — APPROVED** [review]
@rmartinc Nice! Thanks

**Pull request merged**